File: python/problem47.py
import math


def distinctPrimeFactors(targetFactorCount, targetConsecutiveNumCount):
    consecutiveNumCount = 0
    factorCount = 0
    startingNumber = 0
    numToCheck = 10
    foundOne = False

    while (True):

        if isPrime(numToCheck):
            numToCheck += 1
            foundOne = False
            consecutiveNumCount = 0
            continue

        factorCount = calculatePrimeFactors(numToCheck)
        if factorCount == targetFactorCount:
            if not foundOne:
                foundOne = True
                startingNumber = numToCheck
                consecutiveNumCount = 1
            else:
                consecutiveNumCount += 1
                if consecutiveNumCount == targetConsecutiveNumCount:
                    return startingNumber
        else:
            foundOne = False
            consecutiveNumCount = 0
        numToCheck += 1


def calculatePrimeFactors(num):
    factors = []
    factorsCountMap = {}

    for i in range(2, int(math.sqrt(num))+1):

        if num % i == 0:
            numToTest = num // i
            if isPrime(i):
                updateMap(factorsCountMap, i)
            if isPrime(numToTest):
                updateMap(factorsCountMap, numToTest)

            if numToTest == i:
                continue

            while numToTest > 0:
                if numToTest % i == 0:
                    if isPrime(i):
                        updateMap(factorsCountMap, i)
                    numToTest = numToTest // i
                else:
                    break

    for key in factorsCountMap:
        if factorsCountMap[key] == 1:
            factors.append(key)
        else:
            product = 1
            for p in range(0, factorsCountMap[key]):
                product *= key
            factors.append(product)

    return len(factors)


def updateMap(fMap, factor):
    if factor in fMap:
        fMap[factor] = fMap[factor] + 1
    else:
        fMap[factor] = 1


def isPrime(num):
    if num <= 1:
        return False

    if num == 2:
        return True

    if num % 2 == 0:
        return False

    for i in range(3, int(math.sqrt(num)) + 1, 2):
        if num % i == 0:
            return False

    return True


if __name__ == "__main__":
    print(distinctPrimeFactors(3, 3))

File: go/problem47.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(DistinctPrimeFactors(2, 2))
}

func DistinctPrimeFactors(targetFactorNum int, targetConsecutiveCount int) int {

	consecutiveNumCount := 0
	factorCount := 0
	startingNumber := 0
	numToCheck := 10
	foundOne := false

	for true {

		if isPrime(numToCheck) {
			numToCheck++
			foundOne = false
			consecutiveNumCount = 0
			continue
		}

		factorCount = calculatePrimeFactors(numToCheck)
		if factorCount == targetFactorNum {

			if !foundOne {
				// fmt.Println("Initializing consecutive count to 1 for ", numToCheck);
				foundOne = true
				startingNumber = numToCheck
				consecutiveNumCount = 1
			} else {
				// fmt.Println("Incrementing consecutive count by 1 for " + numToCheck);
				consecutiveNumCount++
				if consecutiveNumCount == targetConsecutiveCount {
					// fmt.Println("Returning for " + numToCheck);
					return startingNumber
				}
			}
		} else {
			foundOne = false
			consecutiveNumCount = 0
		}
		numToCheck++
	}

	return -1

}

func calculatePrimeFactors(num int) int {

	factors := make([]int, 0, 10)
	factorsCountMap := make(map[int]int)

	for i := 2; i <= int(math.Sqrt(float64(num))); i++ {

		if num%i == 0 {
			numToTest := num / i
			if isPrime(i) {
				updateMap(factorsCountMap, i)
			}
			if isPrime(numToTest) {
				updateMap(factorsCountMap, numToTest)
			}
			if numToTest == i {
				continue
			}

			for numToTest > 0 {
				if numToTest%i == 0 {
					if isPrime(i) {
						updateMap(factorsCountMap, i)
					}
					numToTest = numToTest / i
				} else {
					break
				}
			}
		}
	}

	for key, value := range factorsCountMap {
		if value == 1 {
			factors = append(factors, key)
		} else {
			product := 1
			for p := 0; p < value; p++ {
				product *= value
			}
			factors = append(factors, product)
		}
	}

	return len(factors)
}

func updateMap(fMap map[int]int, factor int) map[int]int {

	value, ok := fMap[factor]
	if !ok {
		fMap[factor] = 1
	} else {
		fMap[factor] = value + 1
	}
	return fMap
}

func isPrime(num int) bool {

	if num <= 1 {
		return false
	}

	if num == 2 {
		return true
	}

	if num%2 == 0 {
		return false
	}

	for i := 3; i <= int(math.Sqrt(float64(num))); i += 2 {
		if num%i == 0 {
			return false
		}
	}

	return true
}
